File: prototype/php/src/app/Actions/Notifications/Notify.php
<?php

declare(strict_types=1);

namespace App\Actions\Notifications;

use App\Domain\Notifications\NotificationMessage;
use App\Domain\Notifications\RecipientType;
use App\Models\Notification;

final readonly class Notify
{
    public function __invoke(RecipientType $recipient, int $recipientId, NotificationMessage $message): Notification
    {
        $notification = Notification::to($recipient, $recipientId, $message);

        $this->deliverByEmail($notification);

        return $notification;
    }

    /**
     * The prototype delivers to the in-app inbox only. Mail delivery hangs here,
     * behind the same port shape as MagicLinkDelivery.
     */
    private function deliverByEmail(Notification $notification): void {}
}

File: prototype/rails/src/app/actions/notifications/notify.rb
module Notifications
  class Notify
    def call(recipient_type:, recipient_id:, message:)
      notification = Notification.create!(
        { Notification.recipient_column(recipient_type) => recipient_id }
          .merge(subject: message.subject, body: message.body, url: message.url)
      )

      deliver_by_email(notification)

      notification
    end

    private

    # The prototype delivers to the in-app inbox only. Mail hangs off this hook,
    # behind the same port shape as MagicLinkDelivery.
    def deliver_by_email(notification)
    end
  end
end
